### PR TITLE
issue: 1179641 Wrong mtu custome rt tables

### PR DESCRIPTION
**Pull request by rafiw**

### CODE REVIEW

**Comment by rafiw** [comment]
bot:retest

File: src/vma/proto/neighbour.cpp [hunk 1]
```diff
@@ -405,9 +407,22 @@ bool neigh_entry::post_send_udp(iovec * iov, header *h)
...
 	mem_buf_desc_t* p_mem_buf_desc, *tmp = NULL;
 	tx_packet_template_t *p_pkt;
-
+	int mtu = m_p_ring->get_mtu();
 	size_t sz_data_payload = iov->iov_len;
-	size_t max_ip_payload_size = ((m_p_ring->get_mtu() - sizeof(struct iphdr)) & ~0x7);
+	iphdr* p_hdr = &h->m_header.hdr.m_ip_hdr;
+	route_rule_table_key rtk(p_hdr->daddr, p_hdr->saddr, 0);
+	cache_entry_subject<route_rule_table_key, route_val*>* p_ces = NULL;
+	p_ces = g_p_route_table_mgr->get_entry(rtk);
```
Comment (igor-ivanov) [inline src/vma/proto/neighbour.cpp]: @rafiw pay attention that `get_entry` is not thread safe

File: src/vma/dev/rfs_uc_tcp_gro.cpp [hunk 2]
```diff
@@ -47,7 +49,22 @@ rfs_uc_tcp_gro::rfs_uc_tcp_gro(flow_tuple *flow_spec_5t, ring_simple *p_ring, rf
 	m_p_gro_mgr(&(p_ring->m_gro_mgr)), m_b_active(false), m_b_reserved(false)
 {
 	m_n_buf_max = m_p_gro_mgr->get_buf_max();
-	m_n_byte_max = m_p_gro_mgr->get_byte_max() - p_ring->get_mtu(); 
+	int mtu = p_ring->get_mtu();
+	// m_tos is always 0 in VMA
+	route_rule_table_key rtk(flow_spec_5t->get_dst_ip(), flow_spec_5t->get_src_ip(), 0);
+	cache_entry_subject<route_rule_table_key, route_val*>* p_ces = NULL;
+	p_ces = g_p_route_table_mgr->get_entry(rtk);
```
Comment (igor-ivanov) [inline src/vma/dev/rfs_uc_tcp_gro.cpp]: @rafiw pay attention that `get_entry` is not thread safe

File: src/vma/proto/route_table_mgr.cpp [hunk 3]
```diff
@@ -289,7 +288,31 @@ void route_table_mgr::parse_attr(struct rtattr *rt_attribute, route_val *p_val)
 	case RTA_PREFSRC:
 		p_val->set_src_addr(*(in_addr_t *)RTA_DATA(rt_attribute));
 		break;
+	case RTA_METRICS: {
```
Comment (igor-ivanov) [inline src/vma/proto/route_table_mgr.cpp]: consider moving `{` to next line

File: src/vma/proto/route_table_mgr.cpp [hunk 4]
```diff
@@ -289,7 +288,31 @@ void route_table_mgr::parse_attr(struct rtattr *rt_attribute, route_val *p_val)
...
+				break;
+			default:
+				rt_mgr_logdbg("got unexpected METRICS %d %x",
+					type, *(uint32_t *)RTA_DATA(rta));
+				break;
+			}
+			rta = RTA_NEXT(rta, len);
+		}
+	}
+	break;
```
Comment (igor-ivanov) [inline src/vma/proto/route_table_mgr.cpp]: use the same style as https://github.com/Mellanox/libvma/blob/master/src/vma/dev/cq_mgr_mlx5.cpp#L207

File: src/vma/proto/route_val.cpp [hunk 5]
```diff
@@ -131,3 +134,12 @@ void route_val::print_val()
...
 }
+
+void route_val::set_mtu(uint32_t mtu)
+{
+	if (mtu > g_p_net_device_table_mgr->get_max_mtu()) {
+		rt_val_logdbg("route mtu cannot be bigger then max mtu set on devices");
+	} else {
+		m_mtu = mtu;
+	}
+}
```
Comment (igor-ivanov) [inline src/vma/proto/route_val.cpp]: please consider using inline function as
```
inline void set_mtu(uint32_t mtu) {m_mtu = (mtu > g_p_net_device_table_mgr->get_max_mtu() ? g_p_net_device_table_mgr->get_max_mtu() : mtu)}
```

File: src/vma/proto/neighbour.cpp (hunk 1 above)
Comment (rafiw) [reply src/vma/proto/neighbour.cpp]: @igor-ivanov  i added locking
what do you think?

File: src/vma/proto/route_val.cpp (hunk 5 above)
Comment (rafiw) [reply src/vma/proto/route_val.cpp]: its a problem to put it in the h due to compilation issue not sure it's worth it

**Comment by rafiw** [comment]
bot:retest

File: src/vma/dev/rfs_uc_tcp_gro.cpp [hunk 6]
```diff
@@ -47,7 +49,15 @@ rfs_uc_tcp_gro::rfs_uc_tcp_gro(flow_tuple *flow_spec_5t, ring_simple *p_ring, rf
...
 	m_n_buf_max = m_p_gro_mgr->get_buf_max();
-	m_n_byte_max = m_p_gro_mgr->get_byte_max() - p_ring->get_mtu(); 
+	int mtu = p_ring->get_mtu();
+	route_result res;
+	// m_tos is always 0 in VMA
+	g_p_route_table_mgr->route_resolve(route_rule_table_key(flow_spec_5t->get_dst_ip(),
+					flow_spec_5t->get_src_ip(), 0), res);
+	if (res.mtu) {
+		mtu = res.mtu;
+	}
```
Comment (igor-ivanov) [inline src/vma/dev/rfs_uc_tcp_gro.cpp]: do we need `if (res.mtu && (res.mtu < mtu))` in this place?
do we need to use a minimum MTU value for route and device?

File: src/vma/proto/dst_entry.cpp [hunk 7]
```diff
@@ -309,7 +309,11 @@ bool dst_entry::resolve_ring()
 		}
 		if (m_p_ring) {
 			m_max_inline = m_p_ring->get_max_tx_inline();
-			m_max_inline = std::min(m_max_inline, m_p_net_dev_val->get_mtu() + (uint32_t)m_header.m_transport_header_len);
+			int mtu = m_p_net_dev_val->get_mtu();
+			if (m_p_rt_val && m_p_rt_val->get_mtu() > 0 ) {
+				mtu = m_p_rt_val->get_mtu();
+			}
```
Comment (igor-ivanov) [inline src/vma/proto/dst_entry.cpp]: do we need to use a minimum MTU value for route and device?

File: src/vma/proto/dst_entry.cpp [hunk 8]
```diff
@@ -547,7 +551,12 @@ bool dst_entry::prepare_to_send(const int ratelimit_kbps, bool skip_rules, bool
 		set_state(true);
 		if (resolve_net_dev(is_connect)) {
 			set_src_addr();
-			m_max_udp_payload_size = m_p_net_dev_val->get_mtu() - sizeof(struct iphdr);
+			// overwrite mtu from route if exists
+			int mtu = m_p_net_dev_val->get_mtu();
+			if (m_p_rt_val && m_p_rt_val->get_mtu() > 0 ) {
+				mtu = m_p_rt_val->get_mtu();
+			}
```
Comment (igor-ivanov) [inline src/vma/proto/dst_entry.cpp]: do we need to use a minimum MTU value for route and device?

File: src/vma/proto/dst_entry.cpp [hunk 9]
```diff
@@ -648,7 +657,11 @@ void dst_entry::do_ring_migration(lock_base& socket_lock)
 	ring* old_ring = m_p_ring;
 	m_p_ring = new_ring;
 	m_max_inline = m_p_ring->get_max_tx_inline();
-	m_max_inline = std::min(m_max_inline, m_p_net_dev_val->get_mtu() + (uint32_t)m_header.m_transport_header_len);
+	int mtu = m_p_net_dev_val->get_mtu();
+	if (m_p_rt_val && m_p_rt_val->get_mtu() > 0 ) {
+		mtu = m_p_rt_val->get_mtu();
+	}
```
Comment (igor-ivanov) [inline src/vma/proto/dst_entry.cpp]: do we need to use a minimum MTU value for route and device?

File: src/vma/proto/neighbour.cpp [hunk 10]
```diff
@@ -405,9 +407,16 @@ bool neigh_entry::post_send_udp(iovec * iov, header *h)
...
+	route_result res;
+	int mtu = m_p_ring->get_mtu();
 	size_t sz_data_payload = iov->iov_len;
-	size_t max_ip_payload_size = ((m_p_ring->get_mtu() - sizeof(struct iphdr)) & ~0x7);
+	iphdr* p_hdr = &h->m_header.hdr.m_ip_hdr;
+
+	g_p_route_table_mgr->route_resolve(route_rule_table_key(p_hdr->daddr, p_hdr->saddr, 0), res);
+	if (res.mtu) {
+		mtu = res.mtu;
+	}
```
Comment (igor-ivanov) [inline src/vma/proto/neighbour.cpp]: do we need to use a minimum MTU value for route and device?

File: src/vma/proto/route_val.cpp (hunk 5 above)
Comment (igor-ivanov) [reply src/vma/proto/route_val.cpp]: got it

File: src/vma/proto/vma_lwip.cpp [hunk 11]
```diff
@@ -204,13 +204,17 @@ u16_t vma_lwip::vma_ip_route_mtu(ip_addr_t *dest)
...
+	addr.sin_addr.s_addr = res.p_src;
 	net_device_val* ndv = g_p_net_device_table_mgr->get_net_device_val(addr.sin_addr.s_addr);
 	if (ndv) {
 		ifmtu = ndv->get_mtu();
 	}
-
+	// override with mtu route if exist
+	if (res.mtu > 0) {
+		ifmtu = res.mtu;
+	}
```
Comment (igor-ivanov) [inline src/vma/proto/vma_lwip.cpp]: do we need to use a minimum MTU value for route and device?

**Comment by rafiw** [comment]
bot:retest

**Comment by liranoz12** [comment]
Merged #495

**Pull request closed without merging**